### ISSUE TITLE
pat::Jets in MiniAOD are miskeyed

### ISSUE THREAD

**Issue by rappoccio** [cms-sw/cmssw#20346]
@arizzi found a problem in [applySubstructure.py](https://github.com/cms-sw/cmssw/blob/master/PhysicsTools/PatAlgos/python/slimming/applySubstructure_cff.py). The PUPPI inputs are misconfigured between subjets, jets, and packed candidates.

Investigating now...

**Comment by slava77** [comment]
assign reconstruction

**Comment by arizzi** [comment]
seems broken also in 91X

**Comment by rappoccio** [comment]
This is probably broken since we moved to PUPPI.

**Comment by rappoccio** [comment]
Hi, Folks,

I'm running into some trouble here. My current setup is [here](https://github.com/rappoccio/cmssw/tree/checkingptrs_94x)

I'm not sure I'm using the AssociationMap correctly if PUPPI is used. 

From [here](http://cmslxr.fnal.gov/source/PhysicsTools/PatAlgos/plugins/PATPackedCandidateProducer.cc#0389) it seems that the PUPPI candidates are added to the "pf2pc" Association. However, if I look for the daughters from the subjets in the fat jet, I don't see the same collections. 

For the file I'm running (matrix workflow 135.6), the IDs of the collections are:
   * 1:808: particleFlow
   * 1:1150: generalTracks
   * 2:1314: puppi
   * 2:787: packedPFCandidates

The AK8 ungroomed jet daughters are correctly pointing to "puppi", and the subjet daughters are correctly pointing to "packedPFCandidates", but if I use the pc2pf value map [here](https://github.com/rappoccio/cmssw/blob/ea48b0816038e6089c95f9dcf0d81ed7489e8435/PhysicsTools/PatUtils/plugins/JetSubstructurePacker.cc#L106), the output is pointing to "particleFlow", not "puppi":

```
Subjet with pt 405.987, daughters:  2:787,74,1.67285,0.0265511,-2.4655,0.139526
Looking for key: 1:808, 133
Skipping subjet: 2:594, 0
Subjet with pt 69.1917, daughters:  2:787,83,8.09375,-0.05768,-2.53025,0.139526
Looking for key: 1:808, 45
Skipping subjet: 2:594, 1
```



Any ideas?

**Comment by arizzi** [comment]
Hi Sal,

as the name says pc2pf does "packedCandidate to ParticleFlow" you should
use
"pf2pc" that should contain both
particleFlow to packedCandidates
and
puppi to packedCandidates

I'm not sure we fill a pc2puppi map (but it can be added)

if the subjets point to packedPfCands it means they have been already
rekeyed so why do you want to go back to puppi cands? if you want to match
the remaining FatJet daughters you can go the other way (puppi to
packedCands) and compare on the packedCands refs (also given that the goal
is to just have packedCands refs in the end).

ciao
andrea

**Comment by rappoccio** [comment]
Hi, Andrea,

Yes, I could try to go the other way, but this should also work... having PUPPI in the pf2pc Association doesn't seem to be working, though. This seems like something is wrong, regardless of the final solution. In the meantime I can indeed switch to compare all packed candidates. 

Cheers,
Sal

**Comment by arizzi** [comment]
so you are saying also pf2pc is broken? i.e. pf2pc[PuppiRef] doesn't work?

**Comment by rappoccio** [comment]
I will double check.

**Comment by rappoccio** [comment]
OK good suggestion, comparing all packed candidates is much easier. I have pushed a working version and will make a PR shortly.

**Comment by rappoccio** [comment]
PR #20445 addresses this and fixes the issue.

**Comment by slava77** [comment]
+1

fixed in #20445
